### ISSUE TITLE
STAT tables for families which consist of more than two variable fonts

### ISSUE THREAD

**Issue by m4rc1e** [googlei18n/fontmake#436]
I've attempted writing this issue three times so I'm writing up a common scenario I encounter.

Many families hosted on Google Fonts have been drawn using two glyphs sources e.g [Nunito](https://github.com/googlefonts/nunito). One for Roman styles the other for Italic styles. Since both sources are not MM compatible, we'll end up with two variable fonts. Since Fontmake will generate from each glyphs source independently, neither font will reference the other font in their STAT table. We ideally want STAT tables which are similar to the example given in [MS STAT table spec example 5](https://docs.microsoft.com/en-us/typography/opentype/spec/stat#example-5-a-family-comprised-of-a-non-italic-variable-font-plus-an-italic-variable-font). 

I understand that fonttools' varLib is only able to produce a basic STAT table at the moment. However, when it can generate better STAT tables, I feel fontmake should be aware of all the fonts that exist for the given family and produce STAT tables based on this knowledge.

I find this table rather hard to follow. The spec insists that all variable fonts need this table which is fine. However, are the tables which fontmake currently generates good enough? what are the implications if we ship families which consist of more than two fonts and their stat tables don't reference each other?

Please close this if you feel it isn't relevant Fontmake. I've read the other STAT issues and feel this warrants its own discussion.

**Comment by m4rc1e** [comment]
> what are the implications if we ship families which consist of more than two fonts and their stat tables don't reference each other?

I've just tested this using https://github.com/FAlthausen/Vollkorn-Typeface/tree/master/sources

I generated each source using `fontmake -g source.glyphs -o variable` using fontmake 1.5.0.dev0, Mac OS X 10.12.4, Pages v7 (5576).

When installing the fonts using Font Book v7 (302.1), the following issue was raised

![screen shot 2018-05-29 at 14 17 32](https://user-images.githubusercontent.com/7525512/40661240-1e9700d4-634b-11e8-8060-a6c59e960e5b.png)


Inside pages we get:
![screen shot 2018-05-29 at 14 18 51](https://user-images.githubusercontent.com/7525512/40661292-3b00474e-634b-11e8-98b0-409bdbb8c460.png)

The Italic fonts are being flagged as duplicates. Fontmake produced the following STAT tables.


**Roman**
```
<?xml version="1.0" encoding="UTF-8"?>
<ttFont sfntVersion="\x00\x01\x00\x00" ttLibVersion="3.26">

  <STAT>
    <Version value="0x00010002"/>
    <DesignAxisRecordSize value="8"/>
    <!-- DesignAxisCount=1 -->
    <DesignAxisRecord>
      <Axis index="0">
        <AxisTag value="wght"/>
        <AxisNameID value="256"/>  <!-- Weight -->
        <AxisOrdering value="0"/>
      </Axis>
    </DesignAxisRecord>
    <!-- AxisValueCount=0 -->
    <ElidedFallbackNameID value="2"/>  <!-- Regular -->
  </STAT>

</ttFont>
```

**Italic**
```
<?xml version="1.0" encoding="UTF-8"?>
<ttFont sfntVersion="\x00\x01\x00\x00" ttLibVersion="3.26">

  <STAT>
    <Version value="0x00010002"/>
    <DesignAxisRecordSize value="8"/>
    <!-- DesignAxisCount=1 -->
    <DesignAxisRecord>
      <Axis index="0">
        <AxisTag value="wght"/>
        <AxisNameID value="256"/>  <!-- Weight -->
        <AxisOrdering value="0"/>
      </Axis>
    </DesignAxisRecord>
    <!-- AxisValueCount=0 -->
    <ElidedFallbackNameID value="2"/>  <!-- Italic -->
  </STAT>

</ttFont>
```

If I repeat this test using Source Sans Pro VF, https://github.com/adobe-fonts/source-sans-pro/releases/tag/variable-fonts (has STAT tables which reference each other). The Roman and Italic styles are not being flagged as duplicates. 

![screen shot 2018-05-29 at 14 27 24](https://user-images.githubusercontent.com/7525512/40661745-7137e29e-634c-11e8-86af-e90f94baea9d.png)

There are no collisions but there appears to be a style naming issue, https://github.com/adobe-fonts/source-sans-pro/issues/144.

![screen shot 2018-05-29 at 14 31 53](https://user-images.githubusercontent.com/7525512/40662101-51ace32e-634d-11e8-88e1-6f0e5fd757ba.png)

For our purposes, the default fontmake STAT tables are insufficient. I'll ttx hand edit the Vollkorn STATS for the time being. Once I'm happy I have working tables, I'll post my example to https://github.com/fonttools/fonttools/issues/952

**Comment by behdad** [comment]
To fix properly we would need a way to put both families in the same designspace. Maybe we can have axes in designspace that are marked non-interpolatable. Ie. discrete.  Then Italic would be such an axis. varLib then should deduce that all the non-italic faces should be built into one varfont, and all the italic ones in another...  Naming, etc, become more complicated.

We have talked about this before, with @justvanrossum and @letterror

**Comment by davelab6** [comment]
This seems like a good long term solution :) Since Marc is dealing with
this this week, he'll proceed with two separate .glyphs files and a
build.py that hotfixes a new stat table into the TTF outputs. Perhaps that
build.py can reference a fontmake.yml for the specific naming data. But
this is a short term solution.

**Comment by brawer** [comment]
This pending feature request might be related: https://github.com/adobe-type-tools/afdko/issues/176

**Comment by LettError** [comment]
Storing incompatible structures in a single designspace would be very messy. It's not intended for families. Designspace is not going to solve the STAT table one way or the other.

**Comment by justvanrossum** [comment]
Alternatively, fontmake could take additional designspace files for the sole purpose of extracting data for STAT.

**Comment by LettError** [comment]
If you want to store an axis-like ("axish") value that is not relevant for calculating instances or deltas, it could be stored in the designspace.lib:
`designspace.lib['com.google.roboto.settings.axes.italic'] = "1"`

**Comment by madig** [comment]
Consider using https://github.com/daltonmaag/statmake. If people like it enough, we can talk about integrating it into fontmake and maybe even fontTools, in whatever form is best.
